File: examples/python/error_handling.py
from __future__ import annotations

import os
import sys
from collections.abc import Callable
from dataclasses import dataclass

from bisibility import BisibilityApiError, BisibilityClient

EXAMPLE_ID = "python-error-handling"


@dataclass(frozen=True)
class ExpectedApiError:
    code: str
    status: int
    title: str


def required_env(name: str) -> str:
    value = os.environ.get(name, "").strip()
    if not value:
        raise RuntimeError(f"{name} is required.")
    return value


def problem_code(error: BisibilityApiError) -> str | None:
    if error.problem is None:
        return None
    return error.problem.type.rsplit("/", 1)[-1]


def expect_api_error(
    label: str,
    expected: ExpectedApiError,
    operation: Callable[[], object],
) -> None:
    try:
        operation()
    except BisibilityApiError as error:
        if error.status != expected.status:
            raise AssertionError(f"{label} returned status {error.status}.") from error
        if error.problem is None:
            raise AssertionError(f"{label} did not include problem details.") from error
        if error.problem.status != expected.status:
            raise AssertionError(
                f"{label} problem status was {error.problem.status}."
            ) from error
        expected_type = f"https://bisibility.com/problems/{expected.code}"
        if error.problem.type != expected_type:
            raise AssertionError(f"{label} problem type was {error.problem.type}.") from error
        if problem_code(error) != expected.code:
            raise AssertionError(f"{label} problem code was {problem_code(error)}.") from error
        if error.problem.title != expected.title:
            raise AssertionError(f"{label} problem title was {error.problem.title}.") from error
        print(f"{label}: {error.problem.title} ({problem_code(error)})")
        return

    raise RuntimeError(f"{label} did not fail.")


def run() -> None:
    base_url = required_env("BISIBILITY_BASE_URL")
    with (
        BisibilityClient(api_key=required_env("BISIBILITY_API_KEY"), base_url=base_url) as client,
        BisibilityClient(
            api_key=f"{required_env('BISIBILITY_API_KEY')}_invalid",
            base_url=base_url,
        ) as invalid_client,
    ):
        expect_api_error(
            "Invalid API key",
            ExpectedApiError(code="unauthorized", status=401, title="Unauthorized"),
            invalid_client.list_projects,
        )

        expect_api_error(
            "Missing keyword",
            ExpectedApiError(code="not_found", status=404, title="Not found"),
            lambda: client.get_keyword("kw_z00000000000000000000000"),
        )

        projects = client.list_projects()
        project = projects.data[0] if projects.data else None
        if project is None:
            raise RuntimeError("No projects are available for this API key.")

        expect_api_error(
            "Invalid keyword payload",
            ExpectedApiError(code="validation_failed", status=400, title="Validation failed"),
            lambda: client.create_keywords(project.id, {"keywords": [{"keyword": ""}]}),
        )

    print(f"OK {EXAMPLE_ID}")


if __name__ == "__main__":
    try:
        run()
    except BisibilityApiError as error:
        detail = error.problem.detail if error.problem else str(error)
        print(f"API error {error.status}: {detail}", file=sys.stderr)
        raise SystemExit(1) from error
    except Exception as error:
        print(str(error), file=sys.stderr)
        raise SystemExit(1) from error

File: examples/ts/error-handling.ts
import { BisibilityApiError, BisibilityClient } from "@bisibility/sdk";

const exampleId = "ts-error-handling";

function requiredEnv(name: string) {
  const value = process.env[name]?.trim();
  if (!value) {
    throw new Error(`${name} is required.`);
  }
  return value;
}

function assert(condition: unknown, message: string): asserts condition {
  if (!condition) {
    throw new Error(message);
  }
}

function problemCode(error: BisibilityApiError) {
  return error.problem?.type.split("/").at(-1);
}

async function expectApiError(
  label: string,
  expected: { code: string; status: number; title: string },
  operation: () => Promise<unknown>,
) {
  try {
    await operation();
  } catch (error) {
    assert(error instanceof BisibilityApiError, `${label} did not throw BisibilityApiError.`);
    assert(error.status === expected.status, `${label} returned status ${error.status}.`);
    assert(error.problem, `${label} did not include problem details.`);
    assert(
      error.problem.status === expected.status,
      `${label} problem status was ${error.problem.status}.`,
    );
    assert(
      error.problem.type === `https://bisibility.com/problems/${expected.code}`,
      `${label} problem type was ${error.problem.type}.`,
    );
    assert(
      problemCode(error) === expected.code,
      `${label} problem code was ${problemCode(error)}.`,
    );
    assert(
      error.problem.title === expected.title,
      `${label} problem title was ${error.problem.title}.`,
    );
    console.log(`${label}: ${error.problem.title} (${problemCode(error)})`);
    return;
  }

  throw new Error(`${label} did not fail.`);
}

async function run() {
  const baseUrl = requiredEnv("BISIBILITY_BASE_URL");
  const client = new BisibilityClient({
    apiKey: requiredEnv("BISIBILITY_API_KEY"),
    baseUrl,
  });

  const invalidClient = new BisibilityClient({
    apiKey: "bsb_key_test_invalid_error_handling_0001",
    baseUrl,
  });

  await expectApiError(
    "Invalid API key",
    { code: "unauthorized", status: 401, title: "Unauthorized" },
    () => invalidClient.listProjects(),
  );

  await expectApiError(
    "Missing keyword",
    { code: "not_found", status: 404, title: "Not found" },
    () => client.getKeyword("kw_z00000000000000000000000"),
  );

  const projects = await client.listProjects();
  const project = projects.data[0];
  assert(project, "No projects are available for this API key.");

  await expectApiError(
    "Invalid keyword payload",
    { code: "validation_failed", status: 400, title: "Validation failed" },
    () => client.addKeywords(project.id, { keyword: "" }),
  );

  console.log(`OK ${exampleId}`);
}

try {
  await run();
} catch (error: unknown) {
  if (error instanceof BisibilityApiError) {
    console.error(`API error ${error.status}: ${error.problem?.detail ?? error.message}`);
  } else if (error instanceof Error) {
    console.error(error.message);
  } else {
    console.error("Unknown error");
  }
  process.exitCode = 1;
}
